File: week05/19/P4.cpp
#include <iostream>
using namespace std;

struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

ListNode* insert_head(ListNode *head, int val) {
    ListNode *new_node = new ListNode(val, head);
    return new_node;
};

void printList(ListNode* head) {
    ListNode* current_node = head;
    while (current_node != nullptr) {
        cout << current_node->val << " -> ";
        current_node = current_node->next;
    }
    cout << "nullptr" << endl;
}

class Solution {
public:
    ListNode* mergeTwoLists(ListNode* list1, ListNode* list2) {
        ListNode *merge_node = nullptr;
        if (list1 != nullptr && list2 != nullptr) {
            if (list1->val < list2->val) {
                merge_node = list1;
                merge_node->next = mergeTwoLists(list1->next, list2);
            } else {
                merge_node = list2;
                merge_node->next = mergeTwoLists(list1, list2->next);
            }
        }
        else if (list1 != nullptr) merge_node = list1;
        else if (list2 != nullptr) merge_node = list2;
        else merge_node = nullptr;
        return merge_node;
    }
};

int main() {
    ListNode *list1 = nullptr;
    list1 = insert_head(list1, 4);
    list1 = insert_head(list1, 2);
    list1 = insert_head(list1, 1);
    printList(list1);  // 1 -> 2 -> 4 -> nullptr

    ListNode *list2 = nullptr;
    list2 = insert_head(list2, 4);
    list2 = insert_head(list2, 3);
    list2 = insert_head(list2, 1);
    printList(list2);  // 1 -> 3 -> 4 -> nullptr
    
    Solution result;
    ListNode *merge = nullptr;
    merge = result.mergeTwoLists(list1, list2);
    printList(merge);  // 1 -> 1 -> 2 -> 3 -> 4 -> 4 -> nullptr

    return 0;
}

File: week05/19/P4.py
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

def print_list(node):
    curr = node
    while (curr is not None):
        print(curr.val, end=" -> ")
        curr = curr.next
    print("nullptr")

class Solution(object):
    def mergeTwoLists(self, list1, list2):
        if (list1 is None):
            return list2
        elif (list2 is None):
            return list1
        elif (list1.val < list2.val):
            list1.next = self.mergeTwoLists(list1.next, list2)
            return list1
        else:
            list2.next = self.mergeTwoLists(list1, list2.next)
            return list2

if __name__ == "__main__":
    list1 = ListNode(1, ListNode(2, ListNode(4, None)))
    print_list(list1)  # 1 -> 2 -> 4 -> nullptr

    list2 = ListNode(1, ListNode(3, ListNode(4, None)))
    print_list(list2)  # 1 -> 3 -> 4 -> nullptr

    merge = Solution()
    head = merge.mergeTwoLists(list1, list2)
    print_list(head)  # 1 -> 1 -> 2 -> 3 -> 4 -> 4 -> nullptr